### ISSUE TITLE
Project dependencies may have API risk issues

### ISSUE THREAD

**Issue by PyDR-Tool** [eriklindernoren/Keras-GAN#265]
Hi, In **Keras-GAN**, inappropriate dependency versioning constraints can cause risks.

Below are the dependencies and version constraints that the project is using
```
keras*
git+https://www.github.com/keras-team/keras-contrib.git
matplotlib*
numpy*
scipy*
pillow*
scikit-image*
```

The version constraint **==** will introduce the risk of dependency conflicts because the scope of dependencies is too strict.
The version constraint **No Upper Bound** and **\*** will introduce the risk of the missing API Error because the latest version of the dependencies may remove some APIs.

After further analysis, in this project,
The version constraint of dependency **keras** can be changed to *>=0.2.0,<=2.3.1*.
The version constraint of dependency **scipy** can be changed to *>=0.9.0,<=1.7.3*.

The above modification suggestions can reduce the dependency conflicts as much as possible, 
and introduce the latest version as much as possible without calling Error in the projects.

The invocation of the current project includes all the following methods.

<details>
<summary>The calling methods from the keras</summary>
<pre>
cifar10.load_data
mnist.load_data
</pre>
</details>

    
<details>
<summary>The calling methods from the scipy</summary>
<pre>
scipy.ndimage.interpolation.rotate
</pre>
</details>

    
<details>
<summary>The calling methods from the  all methods</summary>
<pre>
model.add
test_accs.append
PixelDA
self.discriminator.compile
y_train.reshape
self.g_AB
self.build_disk_and_q_net
strides.filters.Conv2D
plt.figure
self.build_critic
os.path.exists
self.sample_generator_input
np.random.random
Flatten
imgs.append
self.setup_mnistm
range
batch_size.self.num_classes.np.random.randint.reshape
DiscoGAN
self.G_BA.predict
ACGAN
self.D_A.compile
self.critic.compile
_x2._x1._y2._y1.masked_img.copy
path.scipy.misc.imread.astype
context_encoder.train
y_train.flatten
X_B.reshape
CycleGAN
self.build_vgg
vgg
Pix2Pix
list
open
self.critic_model.train_on_batch
self.d_A.compile
len
self.encoder
l.set_weights
np.arange
enumerate
dropout_rate.Dropout
model
cifar10.load_data
fig.savefig
self.generator_model.compile
infogan.train
K.gradients
sgan.train
d_block
CGAN
plt.imshow
self.build_encoder
np.argmax
f.write
self.G_AB
scipy.misc.imread
imresize
i.imgs.copy
self.sample_images
self.generator
Sequential
np.random.randint
np.random.choice
ccgan.train
Input
self.img_shape.np.prod.self.num_classes.Embedding
RandomWeightedAverage
COGAN
self.sample_interval
self.d_B.train_on_batch
K.log
batch_size.np.random.randint.reshape
acgan.train
filters.Conv2D
to_categorical
self.latent_dim.Dense
np.expand_dims
self.discriminator
os.unlink
K.square
WGANGP
options.open.write
clf_layer
K.sqrt
save
K.random_normal
self.g_AB.predict
self.mnist_y.copy
bigan.train
self.G_BA
LeakyReLU
dcgan.train
zip_f.read
img.copy
self.data_loader.load_data
self.d2
datetime.datetime.now
self.clf
data.read
self.build_generators
self.d1.compile
images.astype
self.build_discriminator
self.img_shape.np.prod.Dense
DCGAN
self.g_BA.predict
np.random.normal
self.D_A.train_on_batch
SRGAN
l.get_weights
self.d_B.compile
self.critic
test_accs.pop
np.repeat
self.generator.predict
np.add
mnist.load_data
j.i.axs.imshow
plt.subplots
self.d2.compile
VGG19
min
self.num_classes.Dense
self.vgg.compile
self.g2
col.row.axs.set_title
self.g1.predict
np.full
Model
scipy.misc.imresize
self.adversarial_autoencoder.train_on_batch
self.d_B
self.save_model
cgan.train
concatenate
ContextEncoder
glob
DUALGAN
bgan.train
np.clip
self.df.Dense
Dense
gzip.GzipFile
self.combined.compile
np.vstack
K.shape
self.build_classifier
gan.train
self.critic.train_on_batch
DataLoader
self.d_A
zip
np.array
i.axs.axis
self.d2.train_on_batch
f_size.filters.Conv2D
print
plt.close
self.save_imgs
deconv2d
self.g2.predict
self.build_decoder
self.mask_randomly
float
self.bigan_generator.compile
Dropout
os.makedirs
np.mean
filepath.replace
Add
X_train.astype
self.decoder.predict
Concatenate
K.exp
self.bigan_generator.train_on_batch
self.latent_dim.self.num_classes.Embedding
self.D_A
RMSprop
urllib.request.urlopen
AdversarialAutoencoder
i.j.axs.imshow
self.g1
model.summary
self.decoder
np.empty
j.i.axs.set_title
UpSampling2D
self.d1
self.generator_model.train_on_batch
SGAN
self.channels.Conv2D
self.vgg
CCGAN
K.random_uniform
self.img_shape.Reshape
self.build_generator
imgs_B.append
i.j.axs.axis
Embedding
aae.train
Adam
self.imread
multiply
self.d1.train_on_batch
np.save
model.to_json
self.discriminator.train_on_batch
imgs_lr.append
j.i.axs.axis
pickle.load
np.concatenate
np.where
np.zeros
d_layer
Activation
self.d_A.train_on_batch
WGAN
self.clf.predict
scipy.ndimage.interpolation.rotate
self.g_BA
gen_imgs.reshape
self.D_B.train_on_batch
partial
ZeroPadding2D
Conv2D
self.setup_mnist
self.G_AB.predict
residual_block
self.encoder.predict
self.build_discriminators
model.save_weights
int
i.axs.imshow
GAN
self.adversarial_autoencoder.compile
BIGAN
LSGAN
np.empty_like
self.normalize
self.D_B
imgs_A.append
wgan.train
np.prod
conv2d
self.D_B.compile
X_A.reshape
col.row.axs.imshow
self.data_loader.load_batch
K.mean
np.load
self.auxilliary
col.row.axs.axis
self.auxilliary.compile
BatchNormalization
imgs_hr.append
BGAN
np.arange.reshape
out_f.write
K.sum
np.fliplr
Reshape
self.critic_model.compile
np.ones
self.combined.train_on_batch
self.vgg.predict
INFOGAN
InstanceNormalization
merge
</pre>
</details>

    

@woctezuma 
Could please help me check this issue?
May I pull a request to fix it?
Thank you very much.
